### ISSUE TITLE
Feature Request: Concurrent Chunk Functionality Adjustment

### ISSUE THREAD

**Issue by 100ent** [FineUploader/fine-uploader#1496]
My api, is set up similar to the the twitter api. Init, Append, Finalize with chunk uploads. So the first hit should be to example.com/api/chunk/init   Second chunk to chunk before last example.com/api/chunk/append ... and so on.

The problem is, when posting the first concurrent upload it always post multiple uploads and not an initial upload. So no actual ajax occurs until the 3rd call to onuploadchunk which by then the setEndpoint is changed to the append address.

This happens regardless of all listeners I've tried: IE onuploadchunk , onuploadchunksuccess and say after first index call set endpoint etc. I've tried other ways as well.

Another feature request I'd love is the ability change the method in a similar fashion as above it would be great to PUT all additional chunks except the first and last. And if any instance option could be changed via callback that would also be amazing but not requesting this just if it's being considered I vote for it.

Anyway, great job guys you have the best jQuery file uploader on the net from what I can see, and I salute and appreciate it.

**Comment by rnicholus** [comment]
I'm afraid I don't understand your first request at all. Perhaps you can attempt to explain a bit differently?

As for your second request, that's not something I plan on implementing. Changing the method depending on the chunk index doesn't make a whole lot of sense to me, and I doubt it would be generally useful.

**Comment by 100ent** [comment]
https://dev.twitter.com/rest/reference/post/media/upload-chunked

**Comment by 100ent** [comment]
If you post a concurrent upload. The initial post is max connections in my case 3. When the actual upload (jxhr) event is fired off, it occurs with the endpoint I set on chunkdata index 1. So even though the initial endpoint is example.com/init fineuploader uses the example.com/append ....

**Comment by rnicholus** [comment]
No, Fine Uploader won't change the endpoint. It will send all requests to the endpoint specified. Or perhaps you are asking for the behavior to be changed such that Fine Uploader follows some sort of endpoint convention based on the type of request? That's what methods are for, so I do t expect to implement that.

**Comment by 100ent** [comment]
CODE EXAMPLE
request: { endpoint: example.com/api/init }
chunking: {enabled: true} 
.....
onuploadchunk {
        if(chunkData.partIndex){ //if greater than 0
	        $("#fine-uploader-gallery").fineUploader('setEndpoint', 'http://example.com/api/append');
	}

**Comment by 100ent** [comment]
Not true, try it

**Comment by 100ent** [comment]
All of the code examples first upload's go to the append endpoint and the init endpoint never get's hit on firefox

**Comment by rnicholus** [comment]
Yes, if you change the endpoint via the API, of course Fine Uploader will respect that change.

**Comment by 100ent** [comment]
My point is the change should only occur at the second part index or greater. But it's affecting partIndex 0

**Comment by rnicholus** [comment]
Sorry, this is really confusing. It sounds like you are now filing a bug instead of a feature request. Is this accurate?

**Comment by 100ent** [comment]
Sorry maybe so, I've never used github

**Comment by 100ent** [comment]
I didn't think it was a bug because being that it's concurrent I figured it somewhat made sense that all of the chunks went out to an individual endpoint, and as of now you can change endpoint per set of chunks, just not per individual chunks

**Comment by 100ent** [comment]
So if it's intended that endpoint can be set on each individual chunk than yes this is a bug

**Comment by rnicholus** [comment]
Ok, I think I understand now. You want more control over when each specific chunk is sent. Not something I plan on adjusting in Fine Uploader. Processes like concurrent chunking are complex enough as it is. I wouldn't suggest changing the endpoint for an upload while that upload is in progress, as the results will be hard to determine.

**Comment by 100ent** [comment]
Well there are many ways I could fix this. One would be the ability to toggle concurrent uploads via API

**Comment by 100ent** [comment]
And a correction I don't want to change the endpoint per each individual chunk personally. My first chunk needs to go to init, just like twitter's api

**Comment by 100ent** [comment]
All middle chunks can go to append, and last (finalize) endpoint is already implemented

**Comment by 100ent** [comment]
So if after my first upload I could turn on concurrent uploads so all the following get sent to a new endpoint that would be fine
EXAMPLE

onchunkupleoad success
     setendpoint : example.com/append
    setconcurrent: true

**Comment by 100ent** [comment]
TWITTER |
INIT	Initialize an upload	command, total_bytes, media_type	JSON
APPEND	Upload(s) of chunked data	command, media_id, segment_index, either media or media_data	HTTP 2XX
FINALIZE	Complete the upload	command, media_id	JSON
STATUS	Check the processing status of the upload (for async video uploads only)	command, media_id	JSON

**Comment by rnicholus** [comment]
You can't currently turn on/off concurrent uploads dynamically via the API. Your questions are a bit all over the place and hard to follow. If you have a good idea as to what code should be changed, feel free to open a pull request and I will take a look.

**Comment by 100ent** [comment]
I don't understand why my questions are all over the place. I am usually considered a great communicator what don't you understand? It's straight forward, setendpoint gets obliterated in concurrent uploads. The only used endpoint is whatever endpoint is set on the last chunk of each concurrent upload. IE upload 3 chunks
chunk1 set endpoint x
chunk2 set endpoint y
chunk3 set endpoint z

All 3 chunks get posted to chunk endpoint z.

This is obviously flawed functionality, or at least should be documented in the docs with a note that set endpoint doesn't work as expected with concurrent uploads.

My issue at this point is being that the twitter api, (twitter's team works tirelessly on making API's that are up to standard) can't even work with fine uploader. That alone seems justification to at least consider adjusting this.

But moreover, it's many logical implementations that would involve at the very least an initial endpoint feature, similar to the success. endpoint feature. I will code it myself if you don't see value in what I'm saying, I just didn't want my adjustment to be incompatible with future updates.

Also, another pointer is this success true thing is kinda not a good idea. For me and probably any developer with their own API. I want my api to be able to work with any choice of jQuery file uploader, so I won't conform it to your choices of wording, etc instead I have had to make many tweaks. And the success true is an easy on to eliminate because success can be based on header response just as easily. Point being a more barebones option may be a good idea in the future so that it can be compatible with updates and still allow developers free rein.

**Comment by 100ent** [comment]
Clarification to barebones option being in addition to the s3, azure, basic options etc.
And I do see value in the success true response for anyone just wishing to slap in a file uploader with ease.

**Comment by rnicholus** [comment]
>Also, another pointer is this success true thing is kinda not a good idea.

It's necessary for cross browser compatibility. The docs cover this in several places, and there are a number of issues explaining this as well. 

>success true is an easy on to eliminate because success can be based on header response just as easily. 

No, it can't. Not for IE9 and older.

**Comment by rnicholus** [comment]
We're going in circles at this point. As mentioned before, if you would like a specific change made to the library, feel free to illustrate with code or a pull request. The specifics of your request are still murky to me. Keep in mind that this library has evolved over years of time, and design choices in place that may not make sense to you are likely frustrating simply do to your lack of knowledge in this domain. The goal for Fine Uploader has been consistency and progressive enhancement, as long as that progressive enhancement does not place a burden on users interfacing with the library. Keep this in mind and you explore changes.
